services/dashboard/overview.go

```diff
@@ -0,0 +1,453 @@
+		if err != nil {
+			return nil, "", err
+		}
+		var resBody pagination.APIResponse[devices.Device]
+		if err := json.NewDecoder(res.Body).Decode(&resBody); err != nil {
+			return nil, "", err
+		}
+		return resBody.Data, resBody.Links.Next, nil
+	}
+	type Marker struct {
+		DeviceID  int64   `json:"device_id"`
+		Label     string  `json:"label"`
+		Latitude  float64 `json:"latitude"`
+		Longitude float64 `json:"longitude"`
+	}
+
+	return func(w http.ResponseWriter, r *http.Request) {
+		sensorGroupID := r.URL.Query().Get("sensorgroup")
+		ws, err := upgrader.Upgrade(w, r, nil)
+		if err != nil {
+			web.HTTPError(w, err)
+			return
+		}
+
+		go func() {
+			var nextCursor string
+			for {
+				// Start fetching pages of devices and stream them to the client
+				devices, cursor, err := getDevicePage(sensorGroupID, nextCursor)
+				if err != nil {
+					ws.Close()
+					return
+				}
+
+				for _, dev := range devices {
+					if dev.Latitude == nil || dev.Longitude == nil {
+						continue
+					}
+					writer, err := ws.NextWriter(websocket.TextMessage)
+					if err != nil {
+						log.Printf("cannot open writer for ws: %v\n", err)
+						continue
+					}
+					defer writer.Close()
+					frame := fmt.Sprintf(`{"device_id": %d, "device_code": "%s", "coordinates": [%f,%f]}`, dev.ID, dev.Code, *dev.Latitude, *dev.Longitude)
+					writer.Write([]byte(frame))
+				}
+				nextCursor = cursor
+				if nextCursor == "" {
+					break
+				}
+			}
+		}()
+	}
+}
+
+func overviewDatastream() http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		var err error
+		var start time.Time
+		var end time.Time
+
+		startQ := r.URL.Query().Get("start")
+		endQ := r.URL.Query().Get("end")
+		if startQ != "" {
+			start, err = time.Parse("2006-01-02", startQ)
+			if err != nil {
+				web.HTTPError(w, err)
+			}
+		}
+		if endQ != "" {
+			end, err = time.Parse("2006-01-02", endQ)
+			if err != nil {
+				web.HTTPError(w, err)
+			}
+		}
+
+		if start.IsZero() {
+			start = time.Now().Add(-7 * 24 * time.Hour)
+		}
+		if end.IsZero() {
+			end = time.Now()
+		}
+		res, err := http.Get(fmt.Sprintf("http://core:3000/datastreams/%s", chi.URLParam(r, "id")))
+		if err != nil {
+			web.HTTPError(w, err)
+			return
+		}
+		var resBody web.APIResponse[coretransport.GetDatastreamResponse]
+		if err := json.NewDecoder(res.Body).Decode(&resBody); err != nil {
+			web.HTTPError(w, err)
+			return
+		}
+
+		page := &views.DatastreamPage{
+			Datastream: *resBody.Data.Datastream,
+			Device:     *resBody.Data.Device,
+			Sensor:     *resBody.Data.Sensor,
+			Start:      start,
+			End:        end,
+		}
+		if isHX(r) {
+			page.WriteBody(w)
+			return
+		}
+
+		views.WriteIndex(w, page)
+	}
+}
+
+func overviewDatastreamStream() http.HandlerFunc {
+	upgrader := websocket.Upgrader{
+		ReadBufferSize:  1024,
+		WriteBufferSize: 1024,
+	}
+	getMeasurementsPage := func(dsID, cursor string) ([]measurements.Measurement, string, error) {
```

Comment:
going to a non-existent link crashes the whole dashboard: http://localhost:3000/overview/datastreams/blablalbla

Reply:
Fixed. Now forwards the APIError properly. The user is shown an understandable error, though its not pretty per se.